src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/TokenCertificateService.java

```diff
@@ -556,29 +561,82 @@ public void registerAuthCert(String hash, String securityServerAddress) throws C
         verifyCertAction(PossibleActionEnum.UNREGISTER, certificateInfo, hash);
         SecurityServerId securityServerId = serverConfService.getSecurityServerId();
         try {
-            managementRequestSenderService.sendAuthCertDeletionRequest(securityServerId,
-                    certificateInfo.getCertificateBytes());
+            if (!skipUnregister) {
+                managementRequestSenderService.sendAuthCertDeletionRequest(securityServerId,
+                        certificateInfo.getCertificateBytes());
+            }
             signerProxyFacade.setCertStatus(certificateInfo.getId(), CertificateInfo.STATUS_DELINPROG);
-        } catch (GlobalConfService.GlobalConfOutdatedException | CodedException e) {
+        } catch (CodedException ce) {
+            if (isCausedBySslAuthFailure(ce)) {
+                throw new NoValidAuthCertificateException(ce);
+            }
+            throw ce;
+        } catch (GlobalConfService.GlobalConfOutdatedException e) {
             throw e;
         } catch (Exception e) {
             throw new RuntimeException("Could not unregister auth cert", e);
         }
     }
 
+    /**
+     * Send the authentication certificate deletion request to central server and set cert status to
+     * {@link CertificateInfo#STATUS_DELINPROG}
+     * @param hash certificate hash
+     * @throws SignCertificateNotSupportedException
+     * @throws ActionNotPossibleException
+     * @throws GlobalConfService.GlobalConfOutdatedException
+     * @throws InvalidCertificateException
+     * @throws KeyNotFoundException
+     * @throws CertificateNotFoundException
+     * @throws NoValidAuthCertificateException
+     */
+    public void unregisterAuthCert(String hash) throws SignCertificateNotSupportedException,
+            ActionNotPossibleException, GlobalConfService.GlobalConfOutdatedException, InvalidCertificateException,
+            KeyNotFoundException, CertificateNotFoundException, NoValidAuthCertificateException {
+        unregisterAuthCertAndMarkForDeletion(hash, false);
+    }
+
+    /**
+     * Set the authentication certificate status to {@link CertificateInfo#STATUS_DELINPROG}
+     * @param hash certificate hash
+     * @throws SignCertificateNotSupportedException
+     * @throws ActionNotPossibleException
+     * @throws GlobalConfService.GlobalConfOutdatedException
+     * @throws InvalidCertificateException
+     * @throws KeyNotFoundException
+     * @throws CertificateNotFoundException
+     */
+    public void markAuthCertForDeletion(String hash) throws SignCertificateNotSupportedException,
+            ActionNotPossibleException, GlobalConfService.GlobalConfOutdatedException, InvalidCertificateException,
+            KeyNotFoundException, CertificateNotFoundException {
+        try {
+            unregisterAuthCertAndMarkForDeletion(hash, true);
+        } catch (NoValidAuthCertificateException e) {
+            // Not a possible exception because the unregister request was never sent
```

Comment:
I think silent catches should still be avoided always, even if we know that exception is impossible currently. Throwing runtime exception is better since at least it protects against mistakes done in future, when e.g. someone modifies unregisterAuthCertAndMarkForDeletion and impossible becomes possible. Also `throw new RuntimeException(e);` is easy to add.

Reply:
I agree that normally they should be avoided but this is somewhat the exception that proves the rule (pardon the pun); I was thinking that skipping the `NoValidAuthCertificateException` is exactly what we want to do here (since this is the `skipUnregister` method). This also hides a redundant exception from the controller layer. Anyway this will probably change with the refactoring that I'm doing to this

Reply:
I am not sure I understand. `catch (NoValidAuthCertificateException e)` should never happen currently, right? We know that by reasoning, but the poor compiler does not know it.

Since it never happens, adding a `throw new RuntimeException(e);` after the comment (I think the comment should stay, it is good to have) does not do anything. 

If instead our reasoning was wrong, or someone later modifies unregisterAuthCertAndMarkForDeletion so that it _does_ throw NoValidAuthCertificateException for some reason (stranger things have happened), we get a RuntimeException instead of http 200 OK. 

NoValidAuthCertificateException is still hidden from the controller layer.

I am guessing Sonar would also complain about current implementation.

So I am suggesting this:

```
        try {
            unregisterAuthCertAndMarkForDeletion(hash, true);
        } catch (NoValidAuthCertificateException e) {
            // Not a possible exception because the unregister request was never sent
            throw new RuntimeException(e);
        }
```

Reply:
It is now `ManagementRequestSendingFailedException` and it is also thrown to the controller layer, where it is wrapped inside `InternalServerErrorException`, because in case of e.g. `IOException` we still want to see the error message in frontend.

But to elaborate more on the previous subject: I was thinking something along the lines that the user is deliberately using an API and a method that is supposed to only mark the cert as `DELETE_IN_PROGRESS` and not send any requests at all -> there is no reason to stop code execution in the case of `NoValidAuthCertificateException` because the cert has been, in fact, successfully marked as `DELETE_IN_PROGRESS`. Of course this probably indicates a flaw in design over anything else and like you said it is probably not ok for SonarQube. 

I'll finish the rest of the task and ask for more comments

Reply:
TokenCertificateService.markAuthCertForDeletion could still remove one exception type which should never happen (but should still result in RuntimeException, for safety):

```
    public void markAuthCertForDeletion(String hash) throws SignCertificateNotSupportedException,
            ActionNotPossibleException, GlobalConfService.GlobalConfOutdatedException, InvalidCertificateException,
            KeyNotFoundException, CertificateNotFoundException {
        try {
            unregisterAuthCertAndMarkForDeletion(hash, true);
        } catch (ManagementRequestSenderService.ManagementRequestSendingFailedException e) {
            // should never happen since we're not sending management requests
            throw new RuntimeException(e);
        }
    }

```